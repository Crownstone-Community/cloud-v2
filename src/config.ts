/**
 * This file will house all used environmental variables we use in the application.
 * It does not cover environmental variables used in dependencies.
 */
export const CONFIG = {
  mongoFilesURL: process.env.FILES_DB_URL ?? "mongodb://127.0.0.1:27017/files_v1",
  mongoDataURL:  process.env.DATA_DB_URL  ?? "mongodb://127.0.0.1:27017/data_v1",
  mongoUserURL:  process.env.USER_DB_URL  ?? "mongodb://127.0.0.1:27017/users_v1",

  emailValidationRequired: process.env.EMAIL_VALIDATION_REQUIRED ?? true,
  generateCustomIds: process.env.GENERATE_CUSTOM_IDS ?? false,

  SSEToken: process.env.SSE_TOKEN || "sse_token"
}

export const SecurityTypes = {
  accessToken: 'AccessToken',
}

