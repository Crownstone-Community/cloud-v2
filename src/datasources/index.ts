export * from './data.datasource';
export * from './users.datasource';
export * from './files.datasource';