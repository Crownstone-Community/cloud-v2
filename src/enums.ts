export const keyTypes = {
    ADMIN_KEY:            "ADMIN_KEY",
    MEMBER_KEY:           "MEMBER_KEY",
    BASIC_KEY:            "BASIC_KEY",
    LOCALIZATION_KEY:     "LOCALIZATION_KEY",
    SERVICE_DATA_KEY:     "SERVICE_DATA_KEY",
    MESH_NETWORK_KEY:     "MESH_NETWORK_KEY",
    MESH_APPLICATION_KEY: "MESH_APPLICATION_KEY",
    MESH_DEVICE_KEY:      "MESH_DEVICE_KEY",
    DEVICE_UART_KEY:      "UART_DEVICE_KEY",
};

export const sphereFeatures : Record<string, SphereFeature_t> = {
    ENERGY_COLLECTION_PERMISSION: "ENERGY_COLLECTION_PERMISSION",
}
