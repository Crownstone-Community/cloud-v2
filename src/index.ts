import {ApplicationConfig, ExpressServer} from './server';
import {PopulateRepositoryContainer} from "./modules/containers/RepoContainer";

export * from './server';

const config = {
  rest: {
    // Use the LB4 application as a route. It should not be listening.
    listenOnStart: false,
  },
};

export async function main(options: ApplicationConfig = {}) {

  const server = new ExpressServer(config);
  await server.boot();
  await server.start();

  await PopulateRepositoryContainer(server.lbApp);

  const port = server.lbApp.restServer.config.port ?? 3000;
  const host = server.lbApp.restServer.config.host ?? 'NO-HOST';

  console.log(`Server is running at ${host}:${port}`);

  return server.lbApp;
}

if (require.main === module) {
  // Run the application
  main({}).catch(err => {
    console.error('Cannot start the application.', err);
    process.exit(1);
  });
}
