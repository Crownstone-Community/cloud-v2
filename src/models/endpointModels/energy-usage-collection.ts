import {Entity, model, property} from '@loopback/repository';

@model()
export class EnergyUsageCollection extends Entity {

  @property({required: true})
  stoneId: string;

  @property({ required: true })
  energyUsage: number;

  @property({type: 'date', required: true})
  timestamp: Date;

}
