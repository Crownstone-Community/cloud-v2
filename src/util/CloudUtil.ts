import * as crypto from "crypto";
const uuid = require('node-uuid');


export const CloudUtil = {
  createKey: function() : string {
    return crypto.randomBytes(16).toString('hex');
  },

  createIBeaconUUID() : string {
    return uuid.v4();
  },

  createToken: function() : string {
    return crypto.randomBytes(32).toString('hex');
  },

  hashPassword(plaintextPassword: string) : string {
    let shasum = crypto.createHash('sha1');
    shasum.update(String(plaintextPassword));
    let hashedPassword = shasum.digest('hex');
    return hashedPassword;
  }
}