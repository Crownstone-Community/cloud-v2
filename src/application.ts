import {BootMixin} from '@loopback/boot';
import {ApplicationConfig} from '@loopback/core';
import {
  RestExplorerBindings,
  RestExplorerComponent,
} from '@loopback/rest-explorer';
import {RepositoryMixin} from '@loopback/repository';
import {RestApplication} from '@loopback/rest';
import {ServiceMixin} from '@loopback/service-proxy';
import path from 'path';
import {CrownstoneSequence} from './sequence';
import {
  AuthenticationComponent,
  registerAuthenticationStrategy
} from '@loopback/authentication';
import {
  AuthorizationComponent,
} from '@loopback/authorization';
import {UserService} from './services';
import {AccessTokenStrategy} from "./security/authentication-strategies/access-token-strategy";
import {SecurityTypes} from "./config";
import {SphereAuthorizationComponent} from "./security/authorization-strategies/authorization";

export interface PackageInfo {
  name: string;
  version: string;
  description: string;
}
const pkg: PackageInfo = require('../package.json');

export class CrownstoneCloud extends BootMixin(ServiceMixin(RepositoryMixin(RestApplication))) {
  constructor(options: ApplicationConfig = {}) {
    let executionPath = __dirname;
    if (options.customPath !== undefined) { executionPath = options.customPath; }
    let customPort = process.env.PORT || 3050;
    if (options.rest && options.rest.port !== undefined) {
      customPort = options.rest.port;
    }

    let customHost = process.env.HOST || '127.0.0.1';
    if (options.rest && options.rest.host !== undefined) {
      customHost = options.rest.host;
    }

    super({...options, rest: { ...options.rest, port: customPort, host: customHost }})

    this.api({
      openapi: '3.0.0',
      info: {title: pkg.name, version: pkg.version},
      paths: {},
      components: {securitySchemes: {
        [SecurityTypes.accessToken]: {
          type: 'apiKey',
          in: 'header',
          name:'Authorization'
        },
      }},
      servers:  [{url: '/api'}],
      security: [{[SecurityTypes.accessToken]: []}],
    });


    this.setUpBindings();
    // Bind authentication component related elements
    this.component(AuthenticationComponent);
    this.component(AuthorizationComponent);
    this.component(SphereAuthorizationComponent);


    // authentication
    registerAuthenticationStrategy(this, AccessTokenStrategy);

    // Set up the custom sequence
    this.sequence(CrownstoneSequence);

    // Set up default home page
    this.static('/', path.join(executionPath, '../public'));

    // Customize @loopback/rest-explorer configuration here
    this.configure(RestExplorerBindings.COMPONENT).to({ path: '/explorer' });
    this.component(RestExplorerComponent);

    this.projectRoot = executionPath;

    // We define both here to allow the testing framework to generate coverage over the ts files.
    this.bootOptions = {
      controllers: {
        dirs: ['controllers'],
        extensions: ['.controller.ts','.controller.js'],
        nested: true,
      },
      repositories: {
        dirs: ['repositories'],
        extensions: ['.repository.ts','.repository.js'],
        nested: true,
      },
      datasources: {
        dirs: ['datasources'],
        extensions: ['.datasource.ts','.datasource.js'],
        nested: true,
      },
      services: {
        dirs: ['services'],
        extensions: ['.service.ts','.service.js'],
        nested: true,
      },
    };
  }

  setUpBindings(): void {
    this.bind("UserService").toClass(UserService);
  }
}
