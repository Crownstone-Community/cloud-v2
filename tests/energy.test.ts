import {CONFIG} from "../src/config";
CONFIG.emailValidationRequired = false;

import {CrownstoneCloud} from "../src/application";
import {Client, createRestAppClient} from '@loopback/testlab';
import {clearTestDatabase, createApp, getRepositories} from "./helpers";
import {createHub, createLocation, createSphere, createStone, createUser} from "./builders/createUserData";
import {CloudUtil} from "../src/util/CloudUtil";
import {auth, getToken, login} from "./rest-helpers/rest.helpers";
import {EnergyDataProcessor} from "../src/modules/energy/EnergyProcessor";
import {EnergyUsageCollection} from "../src/models/endpointModels/energy-usage-collection.model";
// import {EnergyDataProcessor} from "../src/modules/energy/EnergyProcessor";

let app    : CrownstoneCloud;
let client : Client;
let repos = getRepositories();

beforeEach(async () => { await clearTestDatabase(); })
beforeAll(async () => {
  app    = await createApp()
  client = createRestAppClient(app);
})
afterAll(async () => { await app.stop(); })

let dbs;
let admin;
let member;
let guest;
let sphere;
let hub;
let stone,  behaviour,  ability,  abilityProperty;
let stone2, behaviour2, ability2, abilityProperty2;
let stone3, behaviour3, ability3, abilityProperty3;
let location;
let token;

function m(x,a) { return new Date('2022-01-01 01:00:00Z').valueOf() + x*5*60*1000 + a*1000}

function gen(stone: any, value: number, timestamp: number) : Partial<EnergyUsageCollection> {
  return {
    stoneId: stone.id,
    energyUsage: value,
    timestamp: new Date(timestamp)
  }
}
function get(arr: any[], stone, value, timestamp) {
  arr.push(gen(stone, value, timestamp))
}

async function populate() {
  // fill with a bit of data for sync
  dbs = getRepositories();
  admin    = await createUser('admin@test.com',  'test', 0);
  member   = await createUser('member@test.com', 'test', 0);
  guest    = await createUser('guest@test.com',  'test', 0);
  sphere   = await createSphere(admin.id, 'mySphere', 0);
  await dbs.sphereAccess.create({sphereId: sphere.id, userId: member.id, role:'member', sphereAuthorizationToken: CloudUtil.createToken()});
  await dbs.sphereAccess.create({sphereId: sphere.id, userId: guest.id,  role:'guest',  sphereAuthorizationToken: CloudUtil.createToken()});
  hub      = await createHub(sphere.id, 'myHub', 0);
  ({stone, behaviour, ability, abilityProperty} = await createStone(sphere.id, 'stone1', 0));
  ({stone: stone2, behaviour: behaviour2, ability: ability2, abilityProperty: abilityProperty2} = await createStone(sphere.id, 'stone2', 0));
  ({stone: stone3, behaviour: behaviour3, ability: ability3, abilityProperty: abilityProperty3} = await createStone(sphere.id, 'stone3', 0));
  location = await createLocation(sphere.id, 'location', 0);

  stone.locationId = location.id;
  await dbs.stone.update(stone)
  token = await getToken(client, admin);
}

async function prepare() {
  await populate();
  let sphereId = sphere.id;
  // enable collection of energy data.
  await client.post(auth(`/spheres/${sphereId}/energyUsageCollectionPermission?permission=true`))
}

test("test blocking energy loading if permission is not granted", async () => {
  await populate();

  // add some energy data
  let data = [];
  get(data, stone, 1000, m(0,0));
  get(data, stone, 2000, m(1,0));
  await client.post(auth(`/spheres/${sphere.id}/energyUsage`)).send(data)
    .expect(({body})=> {
      expect(body).toHaveProperty('error');
      expect(body?.error).toHaveProperty('message', 'Energy collection is not enabled for this sphere.');
      expect(body?.error).toHaveProperty('statusCode', 403);
    })
});


test("test revoking permissions", async () => {
  await prepare();

  // add some energy data
  let data = [];
  get(data, stone, 1000, m(0,0));
  get(data, stone, 2000, m(1,0));
  await client.post(auth(`/spheres/${sphere.id}/energyUsage`)).send(data)
    .expect(({body})=> {
      expect(body).not.toHaveProperty('error');
    })

  await client.post(auth(`/spheres/${sphere.id}/energyUsageCollectionPermission?permission=false`))
  await client.post(auth(`/spheres/${sphere.id}/energyUsage`)).send(data)
    .expect(({body})=> {
      expect(body).toHaveProperty('error');
      expect(body?.error).toHaveProperty('message', 'Energy collection is not enabled for this sphere.');
      expect(body?.error).toHaveProperty('statusCode', 403);
    })
});


test("test energy loading", async () => {
  await prepare();

  // add some energy data
  let data = [];
  get(data, stone, 1000, m(0,0));
  get(data, stone, 2000, m(1,0));
  await client.post(auth(`/spheres/${sphere.id}/energyUsage`)).send(data)

  let processedPoints = await dbs.stoneEnergyProcessed.find()
  expect(processedPoints.length).toBe(2)
  expect(processedPoints[0].energyUsage).toBe(1000)
  expect(processedPoints[1].energyUsage).toBe(2000)
  expect(await dbs.stoneEnergy.find()).toHaveLength(1)
});



test("check processing energy data without interpolation WITH a gap", async () => {
  await prepare();

  let data = [];
  get(data, stone, 1000, m(0,0));
  get(data, stone, 2000, m(1,0));
  get(data, stone, 3000, m(8,0));

  await client.post(auth(`/spheres/${sphere.id}/energyUsage`)).send(data)

  data = [];
  get(data, stone, 4000, m(9,0));

  await client.post(auth(`/spheres/${sphere.id}/energyUsage`)).send(data)

  let processedPoints = await dbs.stoneEnergyProcessed.find()
  expect(processedPoints.length).toBe(4)
  expect(await dbs.stoneEnergy.find()).toHaveLength(1)
});


test("check processing energy data without interpolation WITH a gap 2", async () => {
  await prepare();

  let data = [];
  get(data, stone, 1000, m(0,0));
  get(data, stone, 2000, m(1,0));
  get(data, stone, 3000, m(8,0));
  get(data, stone, 4000, m(9,0));

  await client.post(auth(`/spheres/${sphere.id}/energyUsage`)).send(data)


  let processedPoints = await dbs.stoneEnergyProcessed.find()
  expect(processedPoints.length).toBe(4)
  expect(await dbs.stoneEnergy.find()).toHaveLength(1)
});


test("check processing energy data in normal situation", async () => {
  await prepare();

  let data = [];
  get(data, stone, 1000,  m(0,1));
  get(data, stone, 2000,  m(1,4));
  get(data, stone, 3000,  m(2,0));
  get(data, stone, 4000,  m(5,4));

  await client.post(auth(`/spheres/${sphere.id}/energyUsage`)).send(data)

  let energyPoints    = await dbs.stoneEnergy.find()
  let processedPoints = await dbs.stoneEnergyProcessed.find()

  expect(processedPoints.length).toBe(5)

  expect(processedPoints[0].energyUsage).toBe(Math.round((1000/303)*299+1000));
  expect(processedPoints[1].energyUsage).toBe(3000);
  expect(processedPoints[2].energyUsage).toBe(Math.round((1000/904)*300 + 3000));
  expect(processedPoints[3].energyUsage).toBe(Math.round((1000/904)*600 + 3000));
  expect(processedPoints[4].energyUsage).toBe(Math.round((1000/904)*900 + 3000));

  for (let i = 0; i < energyPoints.length-1;i++) {
    expect(energyPoints[i].processed).toBe(true);
  }
  expect(energyPoints[energyPoints.length -1].processed).toBe(true);
  expect(await dbs.stoneEnergy.find()).toHaveLength(1)
});

test("check reboot detection and handling", async () => {
  await prepare();

  let data = [];
  get(data, stone, 1000,  m(0,1));
  get(data, stone, 0,     m(2,4));
  get(data, stone, 3000,  m(4,0));

  await client.post(auth(`/spheres/${sphere.id}/energyUsage`)).send(data)


  let processedPoints = await dbs.stoneEnergyProcessed.find()
  expect(processedPoints[0].energyUsage).toBe(1000);
  expect(processedPoints[1].energyUsage).toBe(1000);
  expect(processedPoints[2].energyUsage).toBe(Math.round((3000/596)*296) + 1000); // diff/dt * 1periodDt
  expect(processedPoints[3].energyUsage).toBe(3000 + 1000);
  expect(await dbs.stoneEnergy.find()).toHaveLength(1)
});


test("check large gap", async () => {
  await prepare();

  let data = [];

  get(data, stone, 1000, m(0,1))
  get(data, stone, 2000, m(1,1))
  get(data, stone, 3000, m(9,0))
  get(data, stone, 4000, m(10,0))

  await client.post(auth(`/spheres/${sphere.id}/energyUsage`)).send(data)
  let processedPoints = await dbs.stoneEnergyProcessed.find()
  expect(processedPoints.length).toBe(3)
  expect(await dbs.stoneEnergy.find()).toHaveLength(1)
});


test("check resuming after zero measurement gap", async () => {
  await prepare();

  let data = [];

  get(data, stone, 1000, m(0,0))
  get(data, stone, 0, m(1,1))
  await client.post(auth(`/spheres/${sphere.id}/energyUsage`)).send(data)
  
  data = [];
  get(data, stone, 3000, m(9,0))
  get(data, stone, 4000, m(10,0))
  await client.post(auth(`/spheres/${sphere.id}/energyUsage`)).send(data)

  
  let processedPoints = await dbs.stoneEnergyProcessed.find()
  expect(processedPoints[0].energyUsage).toBe(1000);
  expect(processedPoints[1].energyUsage).toBe(1000);
  expect(processedPoints[2].energyUsage).toBe(3000 + 1000);
  expect(processedPoints[3].energyUsage).toBe(4000 + 1000);
  expect(await dbs.stoneEnergy.find()).toHaveLength(1)
});



test("check double reboot gap", async () => {
  await prepare();

  let data = [];

  get(data, stone, 40000, m(0,0))// 40000
  get(data, stone, 0,     m(1,1))// 40000
  get(data, stone, 3000,  m(2,0))// 43000
  get(data, stone, 4000,  m(3,0))// 44000
  get(data, stone, 20000, m(4,0))// 60000
  get(data, stone, 0,     m(5,1))// 60000
  get(data, stone, 3000,  m(6,0))// 63000
  get(data, stone, 4000,  m(7,0))// 64000
  get(data, stone, 20000, m(8,0))// 80000

  await client.post(auth(`/spheres/${sphere.id}/energyUsage`)).send(data)

  let processedPoints = await dbs.stoneEnergyProcessed.find()
  expect(processedPoints[0].energyUsage).toBe(40000);
  expect(processedPoints[1].energyUsage).toBe(40000);
  expect(processedPoints[2].energyUsage).toBe(43000);
  expect(processedPoints[3].energyUsage).toBe(44000);
  expect(processedPoints[4].energyUsage).toBe(60000);
  expect(processedPoints[5].energyUsage).toBe(60000);
  expect(processedPoints[6].energyUsage).toBe(63000);
  expect(processedPoints[7].energyUsage).toBe(64000);
  expect(processedPoints[8].energyUsage).toBe(80000);
  expect(await dbs.stoneEnergy.find()).toHaveLength(1)
});



test("check double reboot gap with not exactly zero numbers", async () => {
  await prepare();

  let data = [];

  get(data, stone, 40000, m(0,0));   // 40000
  get(data, stone, 200, m(1,0));     // 40000
  get(data, stone, 3000, m(2,0));    // 43000
  get(data, stone, 4000, m(3,0));    // 44000
  get(data, stone, 20000, m(4,0));   // 60000
  get(data, stone, 260, m(5,0));     // 60000
  get(data, stone, 3000, m(6,0));    // 63000
  get(data, stone, 4000, m(7,0));    // 64000
  get(data, stone, 20000, m(8,0));   // 80000

  await client.post(auth(`/spheres/${sphere.id}/energyUsage`)).send(data)

  let processedPoints = await dbs.stoneEnergyProcessed.find()
  expect(processedPoints[0].energyUsage).toBe(40000);
  expect(processedPoints[1].energyUsage).toBe(200   + 40000);
  expect(processedPoints[2].energyUsage).toBe(3000  + 40000);
  expect(processedPoints[3].energyUsage).toBe(4000  + 40000);
  expect(processedPoints[4].energyUsage).toBe(20000 + 40000);
  expect(processedPoints[5].energyUsage).toBe(260   + 20000 + 40000);
  expect(processedPoints[6].energyUsage).toBe(3000  + 20000 + 40000);
  expect(processedPoints[7].energyUsage).toBe(4000  + 20000 + 40000);
  expect(processedPoints[8].energyUsage).toBe(20000 + 20000 + 40000);
  expect(await dbs.stoneEnergy.find()).toHaveLength(1)
});



test("check duplicate handling", async () => {
  await prepare();

  let data = [];

  get(data, stone, 1000, m(1,0))
  get(data, stone, 3000, m(2,0))
  get(data, stone, 3000, m(2,0))
  get(data, stone, 3000, m(2,0))
  get(data, stone, 4000, m(3,0))

  await client.post(auth(`/spheres/${sphere.id}/energyUsage`)).send(data)

  let processedPoints = await dbs.stoneEnergyProcessed.find()
  expect(processedPoints.length).toBe(3);
  expect(await dbs.stoneEnergy.find()).toHaveLength(1)
});



test("check correct handling of small decreases in energy", async () => {
  await prepare();

  let data = [];

  get(data, stone, 100100, m(1,0))
  get(data, stone, 100000, m(3,6))
  get(data, stone, 100000, m(4,6))
  get(data, stone, 100100, m(5,6))

  await client.post(auth(`/spheres/${sphere.id}/energyUsage`)).send(data)

  let processedPoints = await dbs.stoneEnergyProcessed.find();
  expect(processedPoints[0].energyUsage).toBe(100100);
  expect(processedPoints[1].energyUsage).toBe(100100);
  expect(processedPoints[2].energyUsage).toBe(100100);
  expect(processedPoints[3].energyUsage).toBe(100100);
  expect(await dbs.stoneEnergy.find()).toHaveLength(1);
});


test("check correct handling of energyData deletion", async () => {
  await prepare();

  let data = [];

  get(data, stone, 100, m(0,0))
  get(data, stone, 101, m(0,1))
  get(data, stone, 102, m(0,2))
  get(data, stone, 103, m(0,4))
  get(data, stone, 103, m(0,30))

  await client.post(auth(`/spheres/${sphere.id}/energyUsage`)).send(data)

  expect(await dbs.stoneEnergy.find()).toHaveLength(1);
  expect(await dbs.stoneEnergyProcessed.find()).toHaveLength(1);

  data = [];
  get(data, stone, 103, m(1,2))
  await client.post(auth(`/spheres/${sphere.id}/energyUsage`)).send(data)
  expect(await dbs.stoneEnergy.find()).toHaveLength(1);
  expect(await dbs.stoneEnergyProcessed.find()).toHaveLength(2);
});

test("check correct handling of energyData deletion without datapoints on intervals", async () => {
  await prepare();

  let data = [];

  get(data, stone, 101, m(0,1))
  get(data, stone, 102, m(0,2))
  get(data, stone, 103, m(0,4))
  get(data, stone, 103, m(0,30))

  await client.post(auth(`/spheres/${sphere.id}/energyUsage`)).send(data)

  expect(await dbs.stoneEnergy.find()).toHaveLength(4);
  expect(await dbs.stoneEnergyProcessed.find()).toHaveLength(0);
});



test("check correct handling of energyData without datapoints on intervals 2", async () => {
  await prepare();

  let data = [];

  get(data, stone, 101, m(0,0))
  get(data, stone, 102, m(10,0))
  get(data, stone, 103, m(10,4))
  get(data, stone, 103, m(10,30))

  await client.post(auth(`/spheres/${sphere.id}/energyUsage`)).send(data)

  expect(await dbs.stoneEnergy.find()).toHaveLength(1);
  expect(await dbs.stoneEnergyProcessed.find()).toHaveLength(2);
});


test("Aggregation of energy usage: month", async () => {
  await prepare();

  function getDate(i) {
    return new Date(2022,1,1,3,40*i,3)
  }

  let data = [];
  let datapoints = 300;
  for (let i = 0; i < datapoints; i++) {
    get(data, stone, i*10000, getDate(i))
  }

  await client.post(auth(`/spheres/${sphere.id}/energyUsage`)).send(data);

  let processor = new EnergyDataProcessor();
  await processor.processStoneAggregations(sphere.id, stone.id);

  expect(await dbs.stoneEnergyProcessed.find({where:{interval: '5m' }})).toHaveLength(0);
  expect(await dbs.stoneEnergyProcessed.find({where:{interval: '1h' }})).toHaveLength(198);
  expect(await dbs.stoneEnergyProcessed.find({where:{interval: '1d' }})).toHaveLength(8);
  expect(await dbs.stoneEnergyProcessed.find({where:{interval: '1M' }})).toHaveLength(0);
  expect(await dbs.stoneEnergyProcessed.find({where:{interval:'fragment'}})).toHaveLength(1);
}, 10000)



test("check getting of energy data, day, week", async () => {
  await prepare();

  function getDate(i) : Date {
    return new Date(2022,1,1,3,40*i,3)
  }

  let data = [];
  let datapoints = 800;
  // console.log('from', getDate(0).toISOString(), 'to', getDate(datapoints).toISOString())
  for (let i = 0; i < datapoints; i++) {
    get(data, stone, i*10000, getDate(i))
  }

  console.time("Process")
  await client.post(auth(`/spheres/${sphere.id}/energyUsage`)).send(data);
  console.timeEnd("Process")

  let processor = new EnergyDataProcessor();
  await processor.processStoneAggregations(sphere.id, stone.id);


  await client.get(auth(`/spheres/${sphere.id}/energyUsage?date=${ getDate(150).toISOString() }&range=day`)).expect(({body}) => {
    expect(body).toHaveLength(24);
    expect(body[0].interval).toBe('1h');
  });

  await client.get(auth(`/spheres/${sphere.id}/energyUsage?date=${ new Date(2022,1,8).toISOString() }&range=week`)).expect(({body}) => {
    expect(body).toHaveLength(7);
    expect(body[0].interval).toBe('1d');
    for (let i = 0; i < body.length;i++) {
      expect(new Date(body[i].timestamp).getDay()).toBe((i+1)%7);
    }
  });
}, 10000);



test("check getting of energy data, month, year", async () => {
  await prepare();

  function getDate(i) : Date {
    return new Date(2022,-1,1,12*i,40*i,3)
  }

  let data = [];
  let datapoints = 800;
  console.log('from', getDate(0).toISOString(), 'to', getDate(datapoints).toISOString())
  for (let i = 0; i < datapoints; i++) {
    get(data, stone, i*10000, getDate(i))
  }

  console.time("Process")
  await client.post(auth(`/spheres/${sphere.id}/energyUsage`)).send(data);
  console.timeEnd("Process")

  let processor = new EnergyDataProcessor();
  await processor.processStoneAggregations(sphere.id, stone.id);

  await client.get(auth(`/spheres/${sphere.id}/energyUsage?date=${ new Date(2022,1,8).toISOString() }&range=month`)).expect(({body}) => {
    expect(body).toHaveLength(28);
    expect(body[0].interval).toBe('1d');
    expect(body).toMatchSnapshot();
  });
  await client.get(auth(`/spheres/${sphere.id}/energyUsage?date=${ new Date(2022,1,8).toISOString() }&range=year`)).expect(({body}) => {
    expect(body).toHaveLength(12);
    expect(body[0].interval).toBe('1M');
  });
}, 10000);
