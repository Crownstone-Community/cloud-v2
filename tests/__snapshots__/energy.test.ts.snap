// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check getting of energy data, month, year 1`] = `
Array [
  Object {
    "energyUsage": 1174736,
    "interval": "1d",
    "stoneId": "1",
    "timestamp": "2022-01-31T23:00:00.000Z",
  },
  Object {
    "energyUsage": 1193684,
    "interval": "1d",
    "stoneId": "1",
    "timestamp": "2022-02-01T23:00:00.000Z",
  },
  Object {
    "energyUsage": 1212631,
    "interval": "1d",
    "stoneId": "1",
    "timestamp": "2022-02-02T23:00:00.000Z",
  },
  Object {
    "energyUsage": 1231578,
    "interval": "1d",
    "stoneId": "1",
    "timestamp": "2022-02-03T23:00:00.000Z",
  },
  Object {
    "energyUsage": 1250526,
    "interval": "1d",
    "stoneId": "1",
    "timestamp": "2022-02-04T23:00:00.000Z",
  },
  Object {
    "energyUsage": 1269473,
    "interval": "1d",
    "stoneId": "1",
    "timestamp": "2022-02-05T23:00:00.000Z",
  },
  Object {
    "energyUsage": 1288420,
    "interval": "1d",
    "stoneId": "1",
    "timestamp": "2022-02-06T23:00:00.000Z",
  },
  Object {
    "energyUsage": 1307368,
    "interval": "1d",
    "stoneId": "1",
    "timestamp": "2022-02-07T23:00:00.000Z",
  },
  Object {
    "energyUsage": 1326315,
    "interval": "1d",
    "stoneId": "1",
    "timestamp": "2022-02-08T23:00:00.000Z",
  },
  Object {
    "energyUsage": 1345263,
    "interval": "1d",
    "stoneId": "1",
    "timestamp": "2022-02-09T23:00:00.000Z",
  },
  Object {
    "energyUsage": 1364210,
    "interval": "1d",
    "stoneId": "1",
    "timestamp": "2022-02-10T23:00:00.000Z",
  },
  Object {
    "energyUsage": 1383157,
    "interval": "1d",
    "stoneId": "1",
    "timestamp": "2022-02-11T23:00:00.000Z",
  },
  Object {
    "energyUsage": 1402105,
    "interval": "1d",
    "stoneId": "1",
    "timestamp": "2022-02-12T23:00:00.000Z",
  },
  Object {
    "energyUsage": 1421052,
    "interval": "1d",
    "stoneId": "1",
    "timestamp": "2022-02-13T23:00:00.000Z",
  },
  Object {
    "energyUsage": 1439999,
    "interval": "1d",
    "stoneId": "1",
    "timestamp": "2022-02-14T23:00:00.000Z",
  },
  Object {
    "energyUsage": 1458947,
    "interval": "1d",
    "stoneId": "1",
    "timestamp": "2022-02-15T23:00:00.000Z",
  },
  Object {
    "energyUsage": 1477894,
    "interval": "1d",
    "stoneId": "1",
    "timestamp": "2022-02-16T23:00:00.000Z",
  },
  Object {
    "energyUsage": 1496841,
    "interval": "1d",
    "stoneId": "1",
    "timestamp": "2022-02-17T23:00:00.000Z",
  },
  Object {
    "energyUsage": 1515789,
    "interval": "1d",
    "stoneId": "1",
    "timestamp": "2022-02-18T23:00:00.000Z",
  },
  Object {
    "energyUsage": 1534736,
    "interval": "1d",
    "stoneId": "1",
    "timestamp": "2022-02-19T23:00:00.000Z",
  },
  Object {
    "energyUsage": 1553684,
    "interval": "1d",
    "stoneId": "1",
    "timestamp": "2022-02-20T23:00:00.000Z",
  },
  Object {
    "energyUsage": 1572631,
    "interval": "1d",
    "stoneId": "1",
    "timestamp": "2022-02-21T23:00:00.000Z",
  },
  Object {
    "energyUsage": 1591578,
    "interval": "1d",
    "stoneId": "1",
    "timestamp": "2022-02-22T23:00:00.000Z",
  },
  Object {
    "energyUsage": 1610526,
    "interval": "1d",
    "stoneId": "1",
    "timestamp": "2022-02-23T23:00:00.000Z",
  },
  Object {
    "energyUsage": 1629473,
    "interval": "1d",
    "stoneId": "1",
    "timestamp": "2022-02-24T23:00:00.000Z",
  },
  Object {
    "energyUsage": 1648420,
    "interval": "1d",
    "stoneId": "1",
    "timestamp": "2022-02-25T23:00:00.000Z",
  },
  Object {
    "energyUsage": 1667368,
    "interval": "1d",
    "stoneId": "1",
    "timestamp": "2022-02-26T23:00:00.000Z",
  },
  Object {
    "energyUsage": 1686315,
    "interval": "1d",
    "stoneId": "1",
    "timestamp": "2022-02-27T23:00:00.000Z",
  },
]
`;
